app/victoria-metrics/main_test.go

```diff
@@ -0,0 +1,270 @@
+		if f() {
+			return nil
+		}
+	}
+	return fmt.Errorf("timeout")
+}
+
+func tearDown() {
+	vminsert.Stop()
+	vmstorage.Stop()
+	vmselect.Stop()
+	if err := httpserver.Stop(*httpListenAddr); err != nil {
+		log.Fatalf("cannot stop the webservice: %s", err)
+	}
+	os.RemoveAll(storagePath)
+}
+
+func TestInfluxDB(t *testing.T) {
+	t.Parallel()
+	tt := readIn("influxdb", t, fmt.Sprintf("%d", insertionTime.UnixNano()))
+	t.Run("influxdb", func(t *testing.T) {
+		for _, test := range tt {
+			t.Run(test.Name, func(t *testing.T) {
+				t.Parallel()
+				httpWrite(t, "http://127.0.0.1"+testHTTPListenAddr+"/write", test.Data)
+				time.Sleep(testWaitForDataTimeout)
+				data := httpRead(t, "http://127.0.0.1"+testHTTPListenAddr, test.Query)
+				rowContains(t, data, test.Result)
+			})
+		}
+	})
+}
+
+func TestGraphite(t *testing.T) {
+	t.Parallel()
+	tt := readIn("graphite", t, fmt.Sprintf("%d", insertionTime.Unix()))
+	t.Run("graphite", func(t *testing.T) {
+		for _, test := range tt {
+			t.Run(test.Name, func(t *testing.T) {
+				t.Parallel()
+				tcpWrite(t, "127.0.0.1"+testStatsDListenAddr, test.Data)
+				time.Sleep(testWaitForDataTimeout)
+				data := httpRead(t, "http://127.0.0.1"+testHTTPListenAddr, test.Query)
+				rowContains(t, data, test.Result)
+			})
+		}
+	})
+}
+
+func TestOpenTSDB(t *testing.T) {
+	t.Parallel()
+	tt := readIn("opentsdb", t, fmt.Sprintf("%d", insertionTime.Unix()))
+	t.Run("opentsdb", func(t *testing.T) {
+		for _, test := range tt {
+			t.Run(test.Name, func(t *testing.T) {
+				t.Parallel()
+				tcpWrite(t, "127.0.0.1"+testOpenTSDBListenAddr, test.Data)
+				time.Sleep(testWaitForDataTimeout)
+				data := httpRead(t, "http://127.0.0.1"+testHTTPListenAddr, test.Query)
+				rowContains(t, data, test.Result)
+			})
+		}
+	})
+}
+
+func readIn(readFor string, t *testing.T, timeStr string) []test {
+	t.Helper()
+	s := newSuite(t)
+	var tt []test
+	s.noError(filepath.Walk(filepath.Join(testFixturesDir, readFor), func(path string, info os.FileInfo, err error) error {
+		if filepath.Ext(path) != ".json" {
+			return nil
+		}
+		b, err := ioutil.ReadFile(path)
+		s.noError(err)
+		item := test{}
+		s.noError(json.Unmarshal(b, &item))
+		item.Data = strings.Replace(item.Data, "{TIME}", timeStr, 1)
+		tt = append(tt, item)
+		return nil
+	}))
+	if len(tt) == 0 {
+		t.Fatalf("no test found in %s", filepath.Join(testFixturesDir, readFor))
+	}
+	return tt
+}
+
+func httpWrite(t *testing.T, address string, data string) {
+	t.Helper()
+	s := newSuite(t)
+	resp, err := http.Post(address, "", bytes.NewBufferString(data))
+	s.noError(err)
+	s.noError(resp.Body.Close())
+	s.equalInt(resp.StatusCode, 204)
+}
+
+func tcpWrite(t *testing.T, address string, data string) {
+	t.Helper()
+	s := newSuite(t)
+	coon, err := net.Dial("tcp", address)
+	s.noError(err)
+	n, err := coon.Write([]byte(data))
+	s.noError(err)
+	s.equalInt(n, len(data))
+}
+
+func httpRead(t *testing.T, address, query string) []Row {
```

Comment:
No need to pass `address` and `query` as separate params since both of them treated as one concatenated string.